drivers/sms_ser.c

```diff
@@ -0,0 +1,701 @@
+        upsdebugx(LOG_DEBUG, "setting status to LB");
+        status_set("LB");
+    } else if (!DeviceData.upsok) {
+        upsdebugx(LOG_DEBUG, "setting status to RB");
+        status_set("RB");
+    } else if (DeviceData.boost) {
+        upsdebugx(LOG_DEBUG, "setting status to BOOST");
+        status_set("BOOST");
+    } else if (!DeviceData.onbattery) {
+        /* sometimes the flag "onacpower" is not set */
+        upsdebugx(LOG_DEBUG, "setting status to OL");
+        status_set("OL");
+    } else {
+        /* None of these parameters is ON, but we got some response,
+		 * so the device is (administratively) OFF ? */
+        upsdebugx(LOG_DEBUG, "setting status to OFF");
+        status_set("OFF");
+    }
+
+    status_commit();
+    dstate_dataok();
+
+    poll_interval = 5;
+}
+
+void upsdrv_shutdown(void) {
+    /* tell the UPS to shut down, then return - DO NOT SLEEP HERE */
+    int retry;
+
+    /* maybe try to detect the UPS here, but try a shutdown even if
+     * it doesn't respond at first if possible */
+
+    /* replace with a proper shutdown function */
+
+    /* you may have to check the line status since the commands
+     * for toggling power are frequently different for OL vs. OB */
+
+    /* OL: this must power cycle the load if possible */
+
+    /* OB: the load must remain off until the power returns */
+    upsdebugx(2, "upsdrv Shutdown execute");
+
+    for (retry = 1; retry <= MAXTRIES; retry++) {
+        if (sms_instcmd("shutdown.stop", NULL) != STAT_INSTCMD_HANDLED) {
+            continue;
+        }
+
+        if (sms_instcmd("shutdown.return", NULL) != STAT_INSTCMD_HANDLED) {
+            continue;
+        }
+
+        upslogx(LOG_ERR, "Shutting down");
+        set_exit_flag(-2); /* EXIT_SUCCESS */
+        return;
+    }
+
+    upslogx(LOG_ERR, "Shutdown failed!");
+    set_exit_flag(-1);
+}
+
+void upsdrv_help(void) {
+}
+
+/* list flags and values that you want to receive via -x */
+void upsdrv_makevartable(void) {
+    char msg[256];
+
+    upsdebugx(LOG_DEBUG, "upsdrv_makevartable");
+
+    snprintf(msg, sizeof msg, "Set reboot delay, in seconds (default=%d).",
+             DEFAULT_BOOTDELAY);
+    addvar(VAR_VALUE, "rebootdelay", msg);
+}
+
+void upsdrv_initups(void) {
+    char *val;
+
+	upsdebugx(LOG_DEBUG, "upsdrv_initups");
+
+    upsfd = ser_open(device_path);
+    ser_set_speed(upsfd, device_path, B2400);
+
+    if ((val = getval("rebootdelay"))) {
+        int ipv = atoi(val);
+        if (ipv >= 0)
+            bootdelay = (unsigned int)ipv;
+    }
+}
+
+void upsdrv_cleanup(void) {
+    upsdebugx(LOG_DEBUG, "upsdrv_cleanup");
+    /* free(dynamic_mem); */
+    ser_close(upsfd, device_path);
+}
+
+uint8_t sms_prepare_get_status(uint8_t *buffer) {
+    buffer[0] = 'Q';
+    buffer[1] = 255;
+    buffer[2] = 255;
+    buffer[3] = 255;
+    buffer[4] = 255;
+    buffer[5] = (buffer[0] + buffer[1] + buffer[2] + buffer[3] + buffer[4]) * 255;
+    buffer[6] = ENDCHAR;
+
+    return 7;
+}
```

Comment:
Feels a bit weird to see code that just pokes bytes in memory, with no internal (e.g. arg) way of knowing how far it can go. Externally we see that the bufOut is 18 bytes etc., but the method itself has no way of checking if it overflows some allocation.

Also, would it help to set a `buffer[7] = '\0';` to be safe (instead of relying on `memset(0...)` outside)? Is this chain of bytes eventually used like a (zero-terminated) string?

Reply:
In the doc there is no null terminator. its the "\r" the "ENDCHAR" (is strange). But there is a good idea to fixed all [7] index to NULL, i will do it.

Reply:
All itens has fixed size, and calling the "ser_send_buf" we inform the length... so, its safe for me.
buffOut and buffIn has a fixed size too.
all "writes and reads" relying on size, send and receive has a fixed size, and i dont know why, but the "terminator char" is "\r" for this driver. Im just following the documentation in the XML.